### PR TITLE
Player rating logic and display new rating after game over

### LINKED ISSUE #50: Bug in overlay, that shows game end
The overlay that pops up is not covering the chess board.
overlaybug

### PR DESCRIPTION
**Pull request by sagarbakhtar**
Calculate players new rating using [Elo rating system](https://en.m.wikipedia.org/wiki/Elo_rating_system)

Api to calculate new rating
- takes old ratings and winner as parameters and send new rating as response

Display new rating on game over
Screenshot 2019-05-05 at 5 05 16 PM

Resolves #50

### CODE REVIEW

File: client/src/apis/chessGameApi.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+const SERVER_API_URL = `${process.env.REACT_APP_SERVER_URL}/api`;
```
Comment (Prabhakar-Poudel) [inline client/src/apis/chessGameApi.js]: REACT_APP_SERVER_URL -> this has been set in configuration file at config/globalConfig
Please pick it up from there.

File: client/src/components/chess-game/ChessGame.css [hunk 2]
```diff
@@ -4,11 +4,11 @@
...
     background: white;
-    height: 35vw;
-    width: 35vw;
-    border: 12px solid white;
+    height: 90vw;
+    width: 90vw;
+    border: 1px solid white;
     opacity: 0.9;
-    top: 0;
+    top: 2.1rem;
```
Comment (Prabhakar-Poudel) [inline client/src/components/chess-game/ChessGame.css]: Make sure it doesn't break in smaller devices.

File: client/src/apis/chessGameApi.js (hunk 1 above)
Comment (Prabhakar-Poudel) [inline client/src/apis/chessGameApi.js]: REACT_APP_SERVER_URL -> this has been set in configuration file at config/globalConfig Please pick it up from there.

File: client/src/components/chess-game/ChessGame.js [hunk 3]
```diff
@@ -15,14 +16,18 @@ import './ChessGame.css';
...
+
 const INIT_STATE = {
   board: [],
   playerColor: '',
   currentPlayer: WHITE_PLAYER,
   playerOneTime: DEFAULT_GAME_TIME_SECS,
   playerTwoTime: DEFAULT_GAME_TIME_SECS,
   isGameOver: false,
-  gameOverMessage: '',
+  playerOneRating: getRandomNumber(1400, 1500),
```
Comment (Prabhakar-Poudel) [inline client/src/components/chess-game/ChessGame.js]: Will this data be comming from server instead?

File: client/src/components/game-over/GameOver.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    this.state = {
+      isLoading: false,
+      playerOneNewRating: 0,
+      playerTwoNewRating: 0,
+    };
+  }
+
+  componentDidUpdate(prevProps) {
+    const { isGameOver } = this.props;
+    if (isGameOver !== prevProps.isGameOver && isGameOver === true) {
```
Comment (Prabhakar-Poudel) [inline client/src/components/game-over/GameOver.js]: Why this?
doesn't `isGameOver` indicate it all?

File: client/src/components/game-over/GameOver.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    case 'draw':
+      return 'draw';
+    default:
+      return '';
+  }
+};
+
+const ratingDiffString = ratingDiff => (ratingDiff < 0 ? ratingDiff.toString() : `+${ratingDiff}`);
+
+export default class GameOver extends Component {
```
Comment (Prabhakar-Poudel) [inline client/src/components/game-over/GameOver.js]: Could be a PureComponent

File: client/src/components/game-over/GameOver.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  componentDidUpdate(prevProps) {
+    const { isGameOver } = this.props;
+    if (isGameOver !== prevProps.isGameOver && isGameOver === true) {
+      this.getRating();
+    }
+  }
+
+  getRating = () => {
```
Comment (Prabhakar-Poudel) [inline client/src/components/game-over/GameOver.js]: This does not return anything, why get?

File: client/src/components/game-over/GameOver.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    case BLACK_PLAYER:
+      return 'player_two';
+    case 'draw':
+      return 'draw';
+    default:
+      return '';
+  }
+};
+
+const ratingDiffString = ratingDiff => (ratingDiff < 0 ? ratingDiff.toString() : `+${ratingDiff}`);
```
Comment (Prabhakar-Poudel) [inline client/src/components/game-over/GameOver.js]: ratingDiff.toString() -> could be ``${ratingDiff}`` for consistency

File: server/src/controllers/chess-game.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import CGUtil from '../utils/chess-game-util';
+
+exports.getRating = (req, res) => {
```
Comment (Prabhakar-Poudel) [inline server/src/controllers/chess-game.js]: we can use es6 here.

File: server/src/index.js [hunk 9]
```diff
@@ -8,20 +8,21 @@ import db from './config/database';
...
+
+db.connectDb();
 app.use(cors());
 app.use(helmet());
+app.use(bodyParser.urlencoded({ extended: true }));
 app.use(bodyParser.json());
 
 app.use('/auth', auth);
-app.get('/api', (req, res) => res.json({ text: 'Online Chess Game!' }));
+app.use('/api', router);
```
Comment (Prabhakar-Poudel) [inline server/src/index.js]: `/api` is a confusing name for en endpoint. Perhaps `/getRatings` `/getResult` etc could convey more meaning about this endpoint.

File: server/src/router.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import express from 'express';
+import chessGameController from './controllers/chess-game';
+
+const router = express.Router();
+
+router.get('/', (req, res) => {
```
Comment (Prabhakar-Poudel) [inline server/src/router.js]: do we still need this path `/`?

File: server/src/controllers/chess-game.js (hunk 8 above)
Comment (Prabhakar-Poudel) [reply server/src/controllers/chess-game.js]: Also, not sure but controllers does not need to know about req/res objects. It can be a standalone function that returns something wheen asked about something.
And the router can handle req/res part.

File: server/src/index.js (hunk 9 above)
Comment (Prabhakar-Poudel) [reply server/src/index.js]: if the router has somehing simillar as `/something`, we can directly call it as `api.chessgame.com/something` than calling `api.chessgame.com/api/something`. As api is more or less implied with or without the subdomain.

File: server/src/utils/chess-game-util.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
+const calcExpectedScore = (_ratingA, _ratingB) => {
+  const ratingA = parseInt(_ratingA, 10);
+  const ratingB = parseInt(_ratingB, 10);
+  const score = 1 / (1 + (10 ** ((ratingB - ratingA) / 400)));
```
Comment (Prabhakar-Poudel) [inline server/src/utils/chess-game-util.js]: Please consider refactoring this. It is little hard to follow.

File: server/src/utils/chess-game-util.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+  return parseFloat(score.toFixed(2));
+};
+
+const calcScores = (winner, pOneExpScore, pTwoExpScore) => {
+  let pOneScore;
+  let pTwoScore;
+
+  switch (winner) {
+    case 'player_one':
```
Comment (Prabhakar-Poudel) [inline server/src/utils/chess-game-util.js]: This is a hard string constrain b/w client & server. This seems to be an ENUM. Which we don't have the luxury of. Possible solutions could be null-boolean (as we have three states), or a config file (which is not great) as both of them need to keep it anyway. Or may be a frozen object, which may be little better.

Comment (Prabhakar-Poudel) [reply server/src/utils/chess-game-util.js]: This reminds me we also need to freeze our configuration files everywhere.

File: server/src/utils/chess-game-util.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    KFactor = 32;
+  } else if (oldRating <= 2400) {
+    KFactor = 24;
+  } else {
+    KFactor = 16;
+  }
+  return Math.round(oldRating + KFactor * (score - expectedScore));
+};
+
+exports.calcRating = (playerOneRating, playerTwoRating, winner) => {
```
Comment (Prabhakar-Poudel) [inline server/src/utils/chess-game-util.js]: es6

File: server/src/utils/chess-game-util.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+      break;
+    case 'player_two':
+      pOneScore = 0;
+      pTwoScore = 1;
+      break;
+    case 'draw':
+      pOneScore = 0.5;
+      pTwoScore = 0.5;
+      break;
+    default:
```
Comment (Prabhakar-Poudel) [inline server/src/utils/chess-game-util.js]: This case must be reported as there can only be `1/0 0/1 0.5/0.5`, may be it could be checked at higher layer and this is never called if there is 4th case.

File: server/src/utils/chess-game-util.js (hunk 13 above)
Comment (Prabhakar-Poudel) [reply server/src/utils/chess-game-util.js]: does score change when it's a draw? Should it? If not we can simply return old results.

File: server/src/utils/chess-game-util.test.js [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import CGUtil from './chess-game-util';
```
Comment (Prabhakar-Poudel) [inline server/src/utils/chess-game-util.test.js:1]: Please add a test that returns actual numbers that you expect to get when you send some value.
like sending (winner=p1, p1=1400, p2=1400) returns (p1=1450, p2=1350).

**Review by Prabhakar-Poudel — CHANGES_REQUESTED** [review]

File: client/src/components/chess-game/ChessGame.js (hunk 3 above)
Comment (sagarbakhtar) [reply client/src/components/chess-game/ChessGame.js]: Once we have a login implemented, will update this to get data from user details

File: server/src/utils/chess-game-util.js (hunk 13 above)
Comment (sagarbakhtar) [reply server/src/utils/chess-game-util.js]: In the rating system (Elo rating system) which we are using on draw,  rating is changed.

File: server/src/controllers/chess-game.js (hunk 8 above)
Comment (sagarbakhtar) [reply server/src/controllers/chess-game.js]: I read about routes and controllers from [here](https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/routes#Create_the_route-handler_callback_functions), we can change if needed.

File: server/src/index.js (hunk 9 above)
Comment (sagarbakhtar) [reply server/src/index.js]: Makes sense.

File: client/src/components/chess-game/ChessGame.js (hunk 3 above)
Comment (Prabhakar-Poudel) [reply client/src/components/chess-game/ChessGame.js]: Please add a TODO there.

File: client/src/apis/chessGameApi.js [hunk 16]
```diff
@@ -0,0 +1,11 @@
+import Config from '../config/globalConfig';
+
+const getRating = ({ playerOneRating, playerTwoRating, winner }) => {
+  const url = `${Config.serverUrl}/get-rating?playerOneRating=${playerOneRating}&playerTwoRating=${playerTwoRating}&winner=${winner}`;
+  return fetch(url)
```
Comment (Prabhakar-Poudel) [inline client/src/apis/chessGameApi.js]: I think @chirag-shinde has placed "axios" as a dependency.
For consistency we can use axios everywhere or drop axios and use fetch everywhere.

File: client/src/components/chess-board/ChessBoard.js [hunk 17]
```diff
@@ -61,9 +61,9 @@ class ChessBoard extends React.Component {
   }
 
   render() {
-    const { squares, board } = this.props;
+    const { childElementRef, squares, board } = this.props;
     return (
-      <div className="chess-board">
+      <div ref={childElementRef} className="chess-board">
```
Comment (Prabhakar-Poudel) [inline client/src/components/chess-board/ChessBoard.js]: Not sure if it's a fitting name.
is it chessBoardRef?

File: client/src/components/chess-game/ChessGame.js [hunk 18]
```diff
@@ -42,13 +49,13 @@ export default class ChessGame extends React.Component {
   constructor(props) {
     super(props);
     this.state = Object.assign({}, INIT_STATE);
-    this.startGame = this.startGame.bind(this);
-    this.movePiece = this.movePiece.bind(this);
-    this.updateBoardState = this.updateBoardState.bind(this);
-    this.calcPossibleMoves = this.calcPossibleMoves.bind(this);
+    this.chessBoarChildRef = React.createRef();
```
Comment (Prabhakar-Poudel) [inline client/src/components/chess-game/ChessGame.js]: typo

File: client/src/components/chess-game/ChessGame.js [hunk 19]
```diff
@@ -61,18 +68,18 @@ export default class ChessGame extends React.Component {
 
   componentWillUnmount() {
     this.clearTimers();
+    disconnect();
```
Comment (Prabhakar-Poudel) [inline client/src/components/chess-game/ChessGame.js:74]: :+1:

File: client/src/components/chess-game/ChessGame.js [hunk 20]
```diff
@@ -118,7 +125,7 @@ export default class ChessGame extends React.Component {
     }
   };
 
-  startGame() {
+  startGame = () => {
```
Comment (Prabhakar-Poudel) [inline client/src/components/chess-game/ChessGame.js:131]: :+1:

Comment (Prabhakar-Poudel) [reply client/src/components/chess-game/ChessGame.js:131]: Make sure we don't have dangling `bind` statements in constructor, as this syntax removes it's usage.

File: client/src/components/chess-game/ChessGame.js [hunk 21]
```diff
@@ -149,49 +156,52 @@ export default class ChessGame extends React.Component {
...
-            <p>{ gameOverMessage }</p>
-            <button type="button" onClick={this.startGame}>Play Again</button>
-          </div>
+          {isGameOver && (
+            <GameOverOverlay
+              playerOneRating={playerOneRating}
+              playerTwoRating={playerTwoRating}
+              winner={winner}
+              startGame={this.startGame}
+              width={chessBoardWidth}
```
Comment (Prabhakar-Poudel) [inline client/src/components/chess-game/ChessGame.js:200]: why not have width & height 100%?
Do we show anything else in end game beside overlay?

File: package.json [hunk 22]
```diff
@@ -21,8 +21,9 @@
     }
   },
   "scripts": {
-    "test": "yarn test:client",
+    "test": "yarn test:client && yarn test:server",
```
Comment (Prabhakar-Poudel) [inline package.json:24]: :+1:

File: server/src/controllers/chess-game.js [hunk 23]
```diff
@@ -0,0 +1,12 @@
+/* eslint-disable import/prefer-default-export */
+import CGUtil from '../utils/chess-game-util';
+
+const getRating = (req, res) => {
+  const { playerOneRating, playerTwoRating, winner } = req.query;
+  const newRating = CGUtil.calcRating(playerOneRating, playerTwoRating, winner);
+  res.json(newRating);
+};
+
+export default {
```
Comment (Prabhakar-Poudel) [inline server/src/controllers/chess-game.js:9]: why unnamed object export?
Also, this file has 
/* eslint-disable import/prefer-default-export */
and has a default export

**Review by Prabhakar-Poudel — COMMENTED** [review]
LGTM after you have read the comments.

File: client/src/components/chess-game/ChessGame.js (hunk 21 above)
Comment (sagarbakhtar) [reply client/src/components/chess-game/ChessGame.js:200]: To show the overlay, div has position absolute and with width and height 100%, it takes width and height of the viewport, not of the parent element.

File: client/src/apis/chessGameApi.js (hunk 16 above)
Comment (sagarbakhtar) [reply client/src/apis/chessGameApi.js]: Ok, will use "axios" then.

File: server/src/controllers/chess-game.js (hunk 23 above)
Comment (sagarbakhtar) [reply server/src/controllers/chess-game.js:9]: Our linter rules ([details](https://github.com/benmosher/eslint-plugin-import/blob/v2.17.2/docs/rules/prefer-default-export.md)) give an error if there is single export and it is not default.
forgot to remove eslint disable comment, thank for pointing out.
